### ISSUE TITLE
Wrapped lines are numbered in source blocks with very long lines

### ISSUE THREAD

**Issue by robertpanzer** [asciidoctor/asciidoctor#1706]
When having very long lines in a source block that are wrapped, the wrapped lines are also numbered, so that the last lines are not numbered.

When you render this document:
```
= Test
:source-highlighter: coderay

[source,xml,numbered]
----
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
...
</project>
----
```
the project element is wrapped over 4 lines, so that the last 3 lines are not getting any line numbers.

Tried both with coderay and with pygments.

**Comment by mojavelinux** [comment]
This problem is related to line numbering. There are few things going on here. First, the nowrap option is not honored in this scenario. So if you add `%nowrap` to the listing, you still get this problem. Second, if the nowrap option is not honored, unfortunately, I don't see how the line numbers can know about where the wraps occur. In that case, I think you have to switch to `coderay-linenums-mode: inline`. But that introduces a new problem that the wrapped lines bleed into the line number gutter. So you almost have to use `nowrap` when you use line numbers (and we need to fix the first problem).

**Comment by mojavelinux** [comment]
To clarify, the focus of this issue is to get the nowrap option to work when using table-based line numbers. I don't know what else we can do address it. I guess down the road we could switch to using individual table rows for each line number and corresponding line of source code, but that would be a major change to the HTML and stylesheet.

**Comment by mojavelinux** [comment]
We should probably document this as well.

**Comment by mojavelinux** [comment]
I'd even consider automatically enabling nowrap when line numbers are on. That might save people a lot of hassle.

**Comment by Rajind** [comment]
@mojavelinux I think enabling nowrap option automatically will be much better. BTW is there any workaround this until this is fixed.

**Comment by Rajind** [comment]
@mojavelinux Sorry. Figured out what you meant. So this works for now. Hope for a fix in next milestone. :-)
![aaa](https://cloud.githubusercontent.com/assets/5234766/17391098/eaceb956-5a30-11e6-8523-12ea9829778f.png)

![bbb](https://cloud.githubusercontent.com/assets/5234766/17391123/28fdd68a-5a31-11e6-8c28-d1b371597601.png)

**Comment by mojavelinux** [comment]
Correct. One way to fix it is to use the inline linenums mode. The other way to fix it is to apply a patch to the stylesheet that sets the `white-space: pre` on the nested `<pre>` element.

**Comment by owenh000** [comment]
One of the important advantages of line numbering, in my opinion, is that it allows content to be wrapped while still showing the reader where the original line breaks were. With regular wrapping, line breaks are added as required and the user may not be sure where the *real* line breaks actually were (an important detail in some cases). Alternatives include adding icons for where the real line breaks are or disabling wrapping, but neither of these work as nicely as line numbering. Just my thoughts—

Here is the workaround that Dan mentioned in the previous comment, as I understand it. When added to a docinfo.html, the following CSS disables wrapping and enables horizontal scrolling for a syntax-highlighted source block with line numbering enabled. This is currently the best option for me because it allows me to still use the `coderay-linenums-mode: table` mode which allows the reader to copy/paste from the source block without getting all the line numbers also. (As far as I can tell, there is no easy way to get decent line numbers with wrapping, as commented [above](https://github.com/asciidoctor/asciidoctor/issues/1706#issuecomment-237352831)).

```
table.CodeRay { display: block; overflow-x: auto; }
td.code > pre { white-space: pre; }
```

If all line lengths are kept below the wrapped length, the Asciidoctor-PDF `--attribute source-highlighter=rouge` option can be used to also have line numbering for the PDF output. Wrapped lines will still "bleed into the line number gutter" there as mentioned [above](https://github.com/asciidoctor/asciidoctor/issues/1706#issuecomment-237351042).

**Comment by mojavelinux** [comment]
It turns out, there was a bug in the CodeRay stylesheet with linenums in table mode. The nested `<pre>` was not inheriting the `pre` setting. Now it should work like Pygments.

I also documented in the user manual that you must set `*-linenums-mode` to `inline` when wrapping is enabled on preformatted blocks. This is a shortcoming of the source highlighters. The problem needs to be fixed there as this is outside of Asciidoctor's control.
